File: mobile_client/lib/pages/photon_setup_page.dart
import 'package:flutter/material.dart';

import '../api/photon.dart';
import '../main.dart' as app;
import '../services/api_client.dart' show apiErrorMessage;
import '../theme.dart';
import '../widgets/glass.dart';

/// Photon (hosted iMessage) provider setup. Loads the current config from
/// `/api/integrations/photon`, lets the user paste their credentials, and POSTs
/// them back.
///
/// The two SECRET fields (project secret + sidecar token) are write-only: the
/// GET never returns their values, only `*_set` booleans. When a secret is
/// already stored we show a dotted placeholder + a "saved — leave blank to keep"
/// hint, and on save we only send a secret key if the user actually typed one
/// (so a blank field preserves the stored secret).
class PhotonSetupPage extends StatefulWidget {
  const PhotonSetupPage({super.key});

  @override
  State<PhotonSetupPage> createState() => _PhotonSetupPageState();
}

class _PhotonSetupPageState extends State<PhotonSetupPage> {
  final _photon = PhotonApi(app.api);

  final _projectIdCtrl = TextEditingController();
  final _projectSecretCtrl = TextEditingController();
  final _notifyTargetCtrl = TextEditingController();
  final _sidecarUrlCtrl = TextEditingController();
  final _sidecarTokenCtrl = TextEditingController();
  final _allowedUsersCtrl = TextEditingController();

  bool _loading = true;
  bool _busy = false;
  String? _error;

  // Whether the server already has each secret stored (from `*_set`). Drives
  // the "leave blank to keep" hint + lets a blank project-secret satisfy the
  // required rule.
  bool _projectSecretSet = false;
  bool _sidecarTokenSet = false;

  static const String _dots = '••••••••';

  @override
  void initState() {
    super.initState();
    _load();
  }

  @override
  void dispose() {
    _projectIdCtrl.dispose();
    _projectSecretCtrl.dispose();
    _notifyTargetCtrl.dispose();
    _sidecarUrlCtrl.dispose();
    _sidecarTokenCtrl.dispose();
    _allowedUsersCtrl.dispose();
    super.dispose();
  }

  Future<void> _load() async {
    setState(() {
      _loading = true;
      _error = null;
    });
    try {
      final cfg = await _photon.getConfig();
      if (!mounted) return;
      _projectIdCtrl.text = cfg.projectId;
      _notifyTargetCtrl.text = cfg.notifyTarget;
      // Server defaults sidecar_url to http://127.0.0.1:8787; keep whatever it
      // returns so the field is never blank unexpectedly.
      _sidecarUrlCtrl.text = cfg.sidecarUrl;
      _allowedUsersCtrl.text = cfg.allowedUsers;
      setState(() {
        _projectSecretSet = cfg.projectSecretSet;
        _sidecarTokenSet = cfg.sidecarTokenSet;
        _loading = false;
      });
    } catch (e) {
      if (!mounted) return;
      setState(() {
        _error = apiErrorMessage(e);
        _loading = false;
      });
    }
  }

  Future<void> _save() async {
    final projectId = _projectIdCtrl.text.trim();
    final projectSecret = _projectSecretCtrl.text; // secrets sent verbatim
    final sidecarToken = _sidecarTokenCtrl.text;

    // Required: project_id always; project_secret unless one is already stored.
    if (projectId.isEmpty) {
      setState(() => _error = 'Project ID is required');
      return;
    }
    if (projectSecret.isEmpty && !_projectSecretSet) {
      setState(() => _error = 'Project secret is required');
      return;
    }

    setState(() {
      _busy = true;
      _error = null;
    });
    try {
      final configured = await _photon.saveConfig(
        projectId: projectId,
        // Only send secrets the user actually typed — blank keeps the stored
        // value (the GET never round-trips a secret, so sending '' would wipe).
        projectSecret: projectSecret.isEmpty ? null : projectSecret,
        notifyTarget: _notifyTargetCtrl.text.trim(),
        sidecarUrl: _sidecarUrlCtrl.text.trim(),
        sidecarToken: sidecarToken.isEmpty ? null : sidecarToken,
        allowedUsers: _allowedUsersCtrl.text.trim(),
      );
      if (!mounted) return;
      // A typed secret is now stored — flip the flags + clear the inputs so the
      // "saved — leave blank to keep" hint shows on the next edit.
      setState(() {
        if (projectSecret.isNotEmpty) _projectSecretSet = true;
        if (sidecarToken.isNotEmpty) _sidecarTokenSet = true;
        _projectSecretCtrl.clear();
        _sidecarTokenCtrl.clear();
      });
      ScaffoldMessenger.of(context).showSnackBar(SnackBar(
        content: Text(configured
            ? 'Photon connected — iMessage is set up.'
            : 'Saved.'),
      ));
    } catch (e) {
      if (!mounted) return;
      setState(() => _error = apiErrorMessage(e));
    } finally {
      if (mounted) setState(() => _busy = false);
    }
  }

  /// Frosted input decoration — same shape pair_page.dart uses.
  InputDecoration _inputDecoration({required String label, String? hint}) {
    return InputDecoration(
      labelText: label,
      hintText: hint,
      filled: true,
      fillColor: JcTheme.glassFill,
      labelStyle: const TextStyle(color: JcTheme.muted),
      hintStyle: const TextStyle(color: JcTheme.muted),
      border: OutlineInputBorder(
        borderRadius: BorderRadius.circular(14),
        borderSide: const BorderSide(color: JcTheme.glassBorder),
      ),
      enabledBorder: OutlineInputBorder(
        borderRadius: BorderRadius.circular(14),
        borderSide: const BorderSide(color: JcTheme.glassBorder),
      ),
      focusedBorder: OutlineInputBorder(
        borderRadius: BorderRadius.circular(14),
        borderSide: const BorderSide(color: JcTheme.primaryBlue, width: 1.4),
      ),
      contentPadding: const EdgeInsets.symmetric(horizontal: 16, vertical: 14),
    );
  }

  Widget _fieldLabel(String text) => Padding(
        padding: const EdgeInsets.only(bottom: 8, left: 2),
        child: Text(text,
            style: const TextStyle(
                color: JcTheme.text, fontSize: 14, fontWeight: FontWeight.w600)),
      );

  Widget _hint(String text) => Padding(
        padding: const EdgeInsets.only(top: 6, left: 2),
        child: Text(text,
            style: const TextStyle(
                fontSize: 11, color: JcTheme.muted, height: 1.4)),
      );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.transparent,
      extendBodyBehindAppBar: true,
      appBar: glassAppBar(context, title: 'Photon (iMessage)', back: true),
      body: AppBackground(
        child: SafeArea(
          child: _loading
              ? const Center(
                  child: CircularProgressIndicator(color: JcTheme.primaryBlue))
              : SingleChildScrollView(
                  padding:
                      const EdgeInsets.symmetric(horizontal: 24, vertical: 16),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: [
                      const SizedBox(height: 8),
                      const Text(
                        'Connect hosted iMessage',
                        style: TextStyle(
                          fontSize: 22,
                          fontWeight: FontWeight.w800,
                          color: JcTheme.text,
                        ),
                      ),
                      const SizedBox(height: 8),
                      const Text(
                        'Create a project at app.photon.codes, then paste its '
                        'credentials from the project settings page. The Photon '
                        'sidecar must be running on your Mac for messages to send.',
                        style: TextStyle(color: JcTheme.muted, height: 1.5),
                      ),
                      const SizedBox(height: 24),

                      // Project ID (required, plain).
                      _fieldLabel('Project ID'),
                      TextField(
                        controller: _projectIdCtrl,
                        autocorrect: false,
                        decoration: _inputDecoration(
                          label: 'Project ID',
                          hint: 'proj_…',
                        ),
                        style: const TextStyle(color: JcTheme.text),
                      ),
                      const SizedBox(height: 18),

                      // Project secret (required, secret/write-only).
                      _fieldLabel('Project secret'),
                      TextField(
                        controller: _projectSecretCtrl,
                        obscureText: true,
                        autocorrect: false,
                        enableSuggestions: false,
                        decoration: _inputDecoration(
                          label: 'Project secret',
                          hint: _projectSecretSet ? _dots : null,
                        ),
                        style: const TextStyle(color: JcTheme.text),
                      ),
                      if (_projectSecretSet)
                        _hint('Saved — leave blank to keep the stored secret.'),
                      const SizedBox(height: 18),

                      // Notify target (plain).
                      _fieldLabel('Notify target'),
                      TextField(
                        controller: _notifyTargetCtrl,
                        autocorrect: false,
                        keyboardType: TextInputType.emailAddress,
                        decoration: _inputDecoration(
                          label: 'Notify target',
                          hint: 'phone number or email',
                        ),
                        style: const TextStyle(color: JcTheme.text),
                      ),
                      _hint('Where Jarvis sends you iMessages.'),
                      const SizedBox(height: 18),

                      // Sidecar URL (plain, defaults to local sidecar).
                      _fieldLabel('Sidecar URL'),
                      TextField(
                        controller: _sidecarUrlCtrl,
                        autocorrect: false,
                        keyboardType: TextInputType.url,
                        decoration: _inputDecoration(
                          label: 'Sidecar URL',
                          hint: 'http://127.0.0.1:8787',
                        ),
                        style: const TextStyle(color: JcTheme.text),
                      ),
                      _hint('The local Photon sidecar address on your Mac.'),
                      const SizedBox(height: 18),

                      // Sidecar token (secret/write-only).
                      _fieldLabel('Sidecar token'),
                      TextField(
                        controller: _sidecarTokenCtrl,
                        obscureText: true,
                        autocorrect: false,
                        enableSuggestions: false,
                        decoration: _inputDecoration(
                          label: 'Sidecar token',
                          hint: _sidecarTokenSet ? _dots : null,
                        ),
                        style: const TextStyle(color: JcTheme.text),
                      ),
                      if (_sidecarTokenSet)
                        _hint('Saved — leave blank to keep the stored token.'),
                      const SizedBox(height: 18),

                      // Allowed users (plain).
                      _fieldLabel('Allowed users'),
                      TextField(
                        controller: _allowedUsersCtrl,
                        autocorrect: false,
                        decoration: _inputDecoration(
                          label: 'Allowed users',
                          hint: 'comma-separated handles',
                        ),
                        style: const TextStyle(color: JcTheme.text),
                      ),
                      _hint('Who may message Jarvis. Leave blank to allow any.'),

                      // Error banner.
                      if (_error != null)
                        Padding(
                          padding: const EdgeInsets.only(top: 18),
                          child: Text(_error!,
                              style: const TextStyle(color: JcTheme.danger)),
                        ),

                      const SizedBox(height: 26),

                      _BlueButton(
                        label: 'Save',
                        busy: _busy,
                        onPressed: _busy ? null : _save,
                      ),
                      const SizedBox(height: 16),
                    ],
                  ),
                ),
        ),
      ),
    );
  }
}

/// Blue primary CTA pill button — same widget pair_page.dart uses (kept private
/// per-file, matching that file's convention rather than introducing a shared
/// export).
class _BlueButton extends StatelessWidget {
  const _BlueButton({
    required this.label,
    this.onPressed,
    this.busy = false,
  });

  final String label;
  final VoidCallback? onPressed;
  final bool busy;

  @override
  Widget build(BuildContext context) {
    final disabled = onPressed == null && !busy;
    return Opacity(
      opacity: disabled ? 0.5 : 1,
      child: Material(
        color: Colors.transparent,
        borderRadius: BorderRadius.circular(22),
        child: InkWell(
          borderRadius: BorderRadius.circular(22),
          onTap: busy ? null : onPressed,
          child: Container(
            width: double.infinity,
            padding: const EdgeInsets.symmetric(horizontal: 22, vertical: 15),
            decoration: BoxDecoration(
              gradient: blueGradient(),
              borderRadius: BorderRadius.circular(22),
              boxShadow: [
                BoxShadow(
                  color: JcTheme.primaryBlue.withValues(alpha: 0.4),
                  blurRadius: 24,
                  offset: const Offset(0, 6),
                ),
              ],
            ),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                if (busy)
                  const SizedBox(
                    width: 16,
                    height: 16,
                    child: CircularProgressIndicator(
                        strokeWidth: 2.2, color: Colors.white),
                  )
                else
                  const SizedBox.shrink(),
                if (busy) const SizedBox(width: 10),
                Text(
                  label,
                  style: const TextStyle(
                    fontSize: 15,
                    fontWeight: FontWeight.w700,
                    color: Colors.white,
                    letterSpacing: 0.2,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

File: ios_app/JarvisCopilot/Copilot/More/Photon/PhotonSetupPage.swift
import SwiftUI
#if canImport(UIKit)
import UIKit
#endif

/// Photon (hosted iMessage) setup, ported from `pages/photon_setup_page.dart`.
///
/// The two secrets are write-only: the GET never returns their values, only
/// `*_set` flags. When one is stored the field shows a dotted placeholder and a
/// "leave blank to keep" hint, and `PhotonStore.save()` omits a secret the user
/// didn't retype — sending "" would wipe it server-side.
struct PhotonSetupPage: View {
    @State private var store: PhotonStore

    /// See `SettingsPage.init` — a view's `init` isn't main-actor-isolated, so
    /// the store can't be a default argument.
    init(store: PhotonStore? = nil) {
        _store = State(initialValue: store ?? MainActor.assumeIsolated { PhotonStore() })
    }

    var body: some View {
        @Bindable var store = store
        return Group {
            if store.isLoading {
                ProgressView().tint(JcTheme.primaryBlue)
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else {
                ScrollView {
                    VStack(alignment: .leading, spacing: 18) {
                        header
                        PhotonStatusPill(status: store.status)
                            .padding(.bottom, 6)

                        PhotonInputField(label: "Project ID", text: $store.projectID,
                                    hint: "proj_…")
                        PhotonInputField(label: "Project secret", text: $store.projectSecret,
                                    hint: store.projectSecretSet ? PhotonStore.secretPlaceholder : "",
                                    secret: true, note: store.projectSecretHint)
                        PhotonInputField(label: "Notify target", text: $store.notifyTarget,
                                    hint: "phone number or email",
                                    keyboard: .emailAddress,
                                    note: "Where Jarvis sends you iMessages.")
                        PhotonInputField(label: "Sidecar URL", text: $store.sidecarURL,
                                    hint: "http://127.0.0.1:8787",
                                    keyboard: .URL,
                                    note: "The local Photon sidecar address on your Mac.")
                        PhotonInputField(label: "Sidecar token", text: $store.sidecarToken,
                                    hint: store.sidecarTokenSet ? PhotonStore.secretPlaceholder : "",
                                    secret: true, note: store.sidecarTokenHint)
                        PhotonInputField(label: "Allowed users", text: $store.allowedUsers,
                                    hint: "comma-separated handles",
                                    note: "Who may message Jarvis. Leave blank to allow any.")

                        VStack(alignment: .leading, spacing: 6) {
                            Toggle(isOn: $store.allowAll) {
                                Text("Allow all inbound senders")
                                    .font(.system(size: 14, weight: .semibold))
                                    .foregroundStyle(JcTheme.text)
                            }
                            .tint(JcTheme.primaryBlue)
                            Text("Recommended when it's just you texting Jarvis. "
                               + "Restart the gateway to apply.")
                                .font(.system(size: 11))
                                .foregroundStyle(JcTheme.muted)
                                .fixedSize(horizontal: false, vertical: true)
                        }

                        if let message = store.errorMessage {
                            Text(message).font(JcText.body).foregroundStyle(JcTheme.danger)
                                .fixedSize(horizontal: false, vertical: true)
                        }

                        BlueButton("Save", busy: store.isSaving) {
                            Task { await store.save() }
                        }
                        .padding(.top, 8)
                    }
                    .padding(.horizontal, 24)
                    .padding(.vertical, 16)
                }
                .scrollDismissesKeyboard(.interactively)
            }
        }
        .jcScreen("Photon (iMessage)")
        .task { if store.isLoading { store.load() } }
        .moreToast($store.toast)
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Connect hosted iMessage")
                .font(.system(size: 22, weight: .heavy))
                .foregroundStyle(JcTheme.text)
            Text("Create a project at app.photon.codes, then paste its credentials "
               + "from the project settings page. The Photon sidecar must be running "
               + "on your Mac for messages to send.")
                .font(JcText.body)
                .foregroundStyle(JcTheme.muted)
                .fixedSize(horizontal: false, vertical: true)
        }
    }
}

/// One labelled input: bold caption, frosted field, optional muted note under it.
/// Named `PhotonInputField` because `PhotonField` is the MODEL of one server-
/// described field (`PhotonModels.swift`).
struct PhotonInputField: View {
    let label: String
    @Binding var text: String
    var hint: String = ""
    var secret: Bool = false
    var keyboard: UIKeyboardType = .default
    var note: String? = nil

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(label)
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(JcTheme.text)
            Group {
                if secret {
                    SecureField(hint, text: $text)
                } else {
                    TextField(hint, text: $text)
                }
            }
            .keyboardType(keyboard)
            .autocorrectionDisabled()
            .textInputAutocapitalization(.never)
            .jcFieldStyle()
            if let note, !note.isEmpty {
                Text(note)
                    .font(.system(size: 11))
                    .foregroundStyle(JcTheme.muted)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
    }
}

/// The connection/health banner: a tinted icon + label from `PhotonStatus`.
struct PhotonStatusPill: View {
    let status: PhotonStatus

    var body: some View {
        let tint = Color(tone: status.tone)
        HStack(spacing: 10) {
            Image(systemName: status.iconName).font(.system(size: 18)).foregroundStyle(tint)
            Text(status.label)
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(tint)
                .fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 10)
        .background(tint.opacity(0.12), in: RoundedRectangle(cornerRadius: 14, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: 14, style: .continuous)
            .strokeBorder(tint.opacity(0.4), lineWidth: 1))
    }
}
